torchdynamo_poc/TorchDynamo Demo.ipynb

```diff
@@ -0,0 +1,407 @@
+     "data": {
+      "text/plain": [
+       "tensor([[1.0413, 1.9294, 1.4181],\n",
+       "        [1.6091, 0.9600, 1.5658]])"
+      ]
+     },
+     "execution_count": 7,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "foo(example_input)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c4ba2cee",
+   "metadata": {},
+   "source": [
+    "## Graph Breaks\n",
+    "\n",
+    "Graph breaks allow running modules and functions with a mix of code expected to run on the backend and code expected to run at the Python level. This means Torch-MLIR does not have to worry about things like data dependent control flow."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "66d5d4a5",
+   "metadata": {},
+   "source": [
+    "### Example 1: Print statements"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "id": "776c2ab2",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "torchdynamo.reset()\n",
+    "@torchdynamo.optimize(make_torch_mlir_compiler(use_tracing=False, device=\"cpu\", verbose=True))\n",
+    "def foo(a, b):\n",
+    "    print(\"Hello!\")\n",
+    "    return a + b"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "id": "964ddfde",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Hello!\n",
+      "Compiling graph...\n",
+      "torch.fx graph:\n",
+      "graph():\n",
+      "    %arg0_1 : [#users=1] = placeholder[target=arg0_1]\n",
+      "    %arg1_1 : [#users=1] = placeholder[target=arg1_1]\n",
+      "    %add : [#users=1] = call_function[target=torch.ops.aten.add](args = (%arg0_1, %arg1_1), kwargs = {})\n",
+      "    return add\n",
+      "\n",
+      "\n",
+      "torch-mlir backend contract graph:\n",
+      "module attributes {torch.debug_module_name = \"_lambda\"} {\n",
+      "  func.func @forward(%arg0: !torch.vtensor<[2,3],f32>, %arg1: !torch.vtensor<[2,3],f32>) -> !torch.vtensor<[2,3],f32> {\n",
+      "    %int1 = torch.constant.int 1\n",
+      "    %0 = torch.aten.add.Tensor %arg0, %arg1, %int1 : !torch.vtensor<[2,3],f32>, !torch.vtensor<[2,3],f32>, !torch.int -> !torch.vtensor<[2,3],f32>\n",
+      "    return %0 : !torch.vtensor<[2,3],f32>\n",
+      "  }\n",
+      "}\n",
+      "\n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "tensor([[1.5328, 1.4332, 1.1009],\n",
+       "        [1.0809, 0.7956, 0.9636]])"
+      ]
+     },
+     "execution_count": 9,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "foo(torch.rand((2, 3)), torch.rand((2, 3)))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "111e855a",
+   "metadata": {},
+   "source": [
+    "This would not work in the current compilation flow from Torch-MLIR."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a97dc641",
+   "metadata": {},
+   "source": [
+    "### Example 2: Control flow"
```

Comment:
It's a bit tough for me to figure out what this section is trying to illustrate.  I get that control flow can be funky, but I can't tell whether this is illustrating that it works correctly or what.  If the intent is to show that control flow works as expected, perhaps showing a simpler if statement with a couple of non-random inputs that exercise each path would be easier to follow.

Reply:
Thanks for the feedback! This notebook was originally created to live demo during a presentation, so not much thought was put into making it self contained. I'll make a new PR making things more clear.

To answer your question, the two examples illustrate pieces of code that can be run with TorchDynamo + Torch-MLIR but would not work with the current compilation flow from Torch-MLIR, since Torch-MLIR has no support for print statements and data dependent control flow.

Reply:
Ah I see, makes sense.  Thanks!